test/node/object_id_tests.js

```diff
@@ -24,6 +24,192 @@ describe('ObjectId', function () {
+
+  it('should throw an error if object with invalid string id is passed in', function () {
+    var objectInvalidString = {
+      id: 'FFFFFFFFFFFFFFFFFFFFFFFG'
+    };
+    expect(() => new ObjectId(objectInvalidString)).to.throw(TypeError);
+  });
+
+  it('should correctly create ObjectId from object with invalid string id and toHexString method', function () {
+    function newToHexString() {
+      return 'BBBBBBBBBBBBBBBBBBBBBBBB';
+    }
+    var objectInvalidString = {
+      id: 'FFFFFFFFFFFFFFFFFFFFFFFG',
+      toHexString: newToHexString
+    };
+    var buf = Buffer.from('BBBBBBBBBBBBBBBBBBBBBBBB', 'hex');
+    expect(Buffer.from(new ObjectId(objectInvalidString).id)).to.deep.equal(buf);
+  });
+
+  it('should throw an error if object with invalid Buffer id is passed in', function () {
+    var objectInvalidBuffer = {
+      id: Buffer.from([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13])
+    };
+    expect(() => new ObjectId(objectInvalidBuffer)).to.throw(TypeError);
+  });
+
+  it('should correctly create ObjectId from object with invalid Buffer id and toHexString method', function () {
+    function newToHexString() {
+      return 'BBBBBBBBBBBBBBBBBBBBBBBB';
+    }
+    var objectInvalidBuffer = {
+      id: Buffer.from([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]),
+      toHexString: newToHexString
+    };
+    var buf = Buffer.from('BBBBBBBBBBBBBBBBBBBBBBBB', 'hex');
+    expect(Buffer.from(new ObjectId(objectInvalidBuffer).id)).to.deep.equal(buf);
+  });
+
+  const numericIO = [
+    { input: 42, output: 42 },
+    { input: 0x2a, output: 0x2a },
+    { input: 4.2, output: 4 },
+    { input: NaN, output: 0 }
+  ];
+
+  for (const { input, output } of numericIO) {
+    it(`should correctly create ObjectId from ${input} and result in ${output}`, function () {
+      const objId = new ObjectId(input);
+      expect(objId).to.have.property('id');
+      expect(objId.id).to.be.instanceOf(Buffer);
+      expect(objId.id.readUInt32BE(0)).to.equal(output);
+    });
+  }
+
+  it('should correctly create ObjectId undefined or null', function () {
+    const objNull = new ObjectId(null);
+    const objNoArg = new ObjectId();
+    const objUndef = new ObjectId(undefined);
+    expect(objNull.id).to.be.instanceOf(Buffer);
+    expect(objNoArg.id).to.be.instanceOf(Buffer);
+    expect(objUndef.id).to.be.instanceOf(Buffer);
+  });
+
+  it('should throw error if non-12 byte non-24 hex string passed in', function () {
+    expect(() => new ObjectId('FFFFFFFFFFFFFFFFFFFFFFFG')).to.throw(TypeError);
+    expect(() => new ObjectId('thisstringisdefinitelytoolong')).to.throw(TypeError);
+    expect(() => new ObjectId('tooshort')).to.throw(TypeError);
+    expect(() => new ObjectId('101010')).to.throw(TypeError);
+    expect(() => new ObjectId('')).to.throw(TypeError);
+  });
+
+  it('should correctly create ObjectId from 12 byte or 24 hex string', function () {
+    var str1 = 'AAAAAAAAAAAAAAAAAAAAAAAA';
+    var str2 = 'FFFFFFFFFFFFFFFFFFFFFFFF';
```

Comment:
why do we want 2 instances of the 24 string? also, the expect statements here are not executing, for the same reason as the other one above (if you use the chaining keywords, like `expect(...).to.equal(...)`, you're less likely to run into this issue)